### ISSUE TITLE
os.environ["CUDA_VISIBLE_DEVICES"] = "" does not use CPU

### ISSUE THREAD

**Issue by ShuhuaGao** [catalyst-team/catalyst#1420]
## 🐛 Bug Report

Even we set `os.environ["CUDA_VISIBLE_DEVICES"] =""`, the model is still trained on the GPU. This contradicts to the [documentation](https://catalyst-team.github.io/catalyst/faq/dp.html).

### How To Reproduce
I simply copy the linear regression minimal example [here](https://github.com/catalyst-team/catalyst#minimal-examples)

 

#### Code sample
```python
# I added these first two lines.
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""


import torch
from torch.utils.data import DataLoader, TensorDataset
from catalyst import dl


# data
num_samples, num_features = int(1e4), int(1e1)
X, y = torch.rand(num_samples, num_features), torch.rand(num_samples)
dataset = TensorDataset(X, y)
loader = DataLoader(dataset, batch_size=32, num_workers=1)
loaders = {"train": loader, "valid": loader}

# model, criterion, optimizer, scheduler
model = torch.nn.Linear(num_features, 1)
criterion = torch.nn.MSELoss()
optimizer = torch.optim.Adam(model.parameters())
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, [3, 6])

# model training
runner = dl.SupervisedRunner()
runner.train(
    model=model,
    criterion=criterion,
    optimizer=optimizer,
    scheduler=scheduler,
    loaders=loaders,
    logdir="./logdir",
    valid_loader="valid",
    valid_metric="loss",
    minimize_valid_metric=True,
    num_epochs=4,
    verbose=True
)

# check model on GPU or CPU
print('Is model on GPU? ', next(model.parameters()).is_cuda)
```

#### Screenshots
Ouput
```
Is model on GPU?  True
```
#### More info
Since the `loader` is on CPU, calling `runner.predict_batch(next(iter(loader)))` leads to the error below:
```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument mat1 in method wrapper_addmm)
```

### Environment
Please copy and paste the output from our environment collection script
```bash
catalyst-contrib --collect-env
# or manually
wget https://raw.githubusercontent.com/catalyst-team/catalyst/master/catalyst/contrib/scripts/collect_env.py
python collect_env.py
```
(or fill out the checklist below manually).

```bash
# example checklist, fill with your info
Catalyst version: 20.04.
PyTorch version: 1.11.0
Python version: 3.9
CUDA runtime version: 11.4
Nvidia driver version: 472.39
cuDNN version: No CUDA
```

### Additional context
If we set `cpu=True` in `runner.train`, then the model indeed lies on the CPU.

**Comment by Scitator** [comment]
If you are running this example into the simple notebook cell - please try to move
```python
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
```
into separate one.

Otherwise, please try [CPUEngine](https://github.com/catalyst-team/catalyst/blob/master/catalyst/engines/torch.py#L20) during train/inference.

**Comment by ShuhuaGao** [comment]
Moving to the first separate cell in Jupyter does not help. The result is still:
```python
print('Is model on GPU? ', next(model.parameters()).is_cuda)
print(runner.engine)
```
```
Is model on GPU?  True
<catalyst.engines.torch.GPUEngine object at 0x000001EF659B1D60>
```

**Comment by Scitator** [comment]
what about CPUEngine?

**Comment by ShuhuaGao** [comment]
@Scitator . Thanks. I figured it out. 


> Moving to the first separate cell in Jupyter does not help

It was because I did not restart the kernel (see [StackOverflow](https://stackoverflow.com/questions/67311527/how-to-set-gpu-count-to-0-using-os-environcuda-visible-devices)).

So the keys are 
- Set `os.environ["CUDA_VISIBLE_DEVICES"]` before `import torch` or any other import that may bring in `torch`. A safe way is to put it in the first cell.
- If `os.environ["CUDA_VISIBLE_DEVICES"]` is reset, e.g., switching from CPU to GPU, we need to **restart** the Jupyter kernel.
